### PR TITLE
Decoupling go-service into 2 parts

### PR DESCRIPTION
**Pull request by gburanov**
The reason for this PR is here
https://github.com/remerge/audience_sync/pull/1#discussion_r176643318

I started to do the splitting, and also found a way to simplify the interface. Some information on how to use is in readme.

## Usage
* Every service should
1. Implement service interface
2. Initialise service executor (ServiceExecutor or ServerlessServiceExecutor)
3. ServerlessServiceExecutor is thinner (does not have Tracker and Server)
4. ServiceExecutor can be used in an implicit composition

Advantages over old solution:
* Adding shutdown is mandatory since it is part of interface now
* Can choose (and switch) from ServiceExecutor to ServerlessServiceExecutor
* Does not need to implement business logic that should be part of go-service

https://github.com/remerge/bloom/blob/00ff954ebf18b9a2e0cc0fed2e3ff7aec7c8d975/main/bloom.go#L13

The PR is not finished, I can polish but do you think the separation in this way make sense? 
@dynamix  @hollow  @itszootime 

At least for my new service it would be valuable

### CODE REVIEW

**Comment by gburanov** [comment]
This is still the case. This PR has 2 goals

- Split service into 2 parts. This is done exactly as u said, "base is the service without a Server and the extension of the base has a Server (and tracker)"

- By creating the Service interface, remove some business logic code from real servers
https://github.com/remerge/trigger/blob/master/main/trigger.go#L12-L15

File: shutdown.go [hunk 1]
```diff
@@ -26,7 +24,13 @@ func WaitForShutdown(timeout time.Duration, handler ShutdownHandler) {
 		syscall.SIGTERM,
 	)
 
-	sig := <-ch
+	var signal os.Signal
+	select {
+	case sig := <-ch:
+		signal = sig
+	case <-done:
```
Comment (gburanov) [inline shutdown.go]: I added the case when the service exit the run() without error. This is probably unusual, but also a valid case.

**Comment by gburanov** [comment]
@dynamix  I change the naming as we discussed. Plz take a llok

**Comment by akaspin** [comment]
Just two cents. I implemented https://github.com/akaspin/supervisor for correct up/down flow. Maybe it be useful for you.

**Comment by itszootime** [comment]
Looks great, but the `S` suffix for `tracker` and `server` structs is a bit cryptic

**Pull request merged**